File: frontend/src/api/blueprints.ts
import { API_BASE_URL } from './config';

export type Blueprint = {
  id: number;
  game_id: number;
  creator_id: number;
  title: string;
  description: string;
  total_levels: number;
  estimated_duration_minutes: number;
  topic_order: number[];
  difficulty_distribution: Record<string, number>;
  reward_style: string;
  branching_style: string;
  approved: boolean;
  created_at: string;
  updated_at: string;
};

export type BlueprintUpdate = Partial<
  Pick<
    Blueprint,
    | 'title'
    | 'description'
    | 'total_levels'
    | 'estimated_duration_minutes'
    | 'topic_order'
    | 'difficulty_distribution'
    | 'reward_style'
    | 'branching_style'
  >
>;

export async function createBlueprint(token: string, gameId: number): Promise<Blueprint> {
  return request<Blueprint>('/api/ai/create-blueprint', {
    method: 'POST',
    token,
    body: JSON.stringify({ game_id: gameId, replace_existing: true }),
  });
}

export async function getBlueprint(token: string, gameId: number): Promise<Blueprint> {
  return request<Blueprint>(`/api/games/${gameId}/blueprint`, { token });
}

export async function updateBlueprint(token: string, gameId: number, payload: BlueprintUpdate): Promise<Blueprint> {
  return request<Blueprint>(`/api/games/${gameId}/blueprint`, {
    method: 'PATCH',
    token,
    body: JSON.stringify(payload),
  });
}

export async function approveBlueprint(token: string, gameId: number): Promise<Blueprint> {
  return request<Blueprint>(`/api/games/${gameId}/blueprint/approve`, {
    method: 'POST',
    token,
  });
}

async function request<T>(path: string, options: RequestInit & { token: string }): Promise<T> {
  const response = await fetch(`${API_BASE_URL}${path}`, {
    ...options,
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${options.token}`,
      ...options.headers,
    },
  });

  if (!response.ok) {
    let message = 'Something went wrong.';
    try {
      const payload = await response.json();
      message = payload.detail ?? message;
    } catch {
      message = response.statusText || message;
    }
    throw new Error(message);
  }

  return response.json() as Promise<T>;
}

File: backend/app/api/routes/blueprints.py
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.api.deps import require_user
from app.db.session import get_db
from app.models.blueprint import GameBlueprint
from app.models.game import Game
from app.models.topic import Topic
from app.models.user import User
from app.schemas.blueprint import BlueprintRead, BlueprintUpdate, CreateBlueprintRequest
from app.services.ai.blueprint_generation import create_blueprint_from_topics

ai_router = APIRouter(prefix="/api/ai", tags=["ai"])
blueprints_router = APIRouter(prefix="/api/games/{game_id}/blueprint", tags=["blueprints"])


@ai_router.post("/create-blueprint", response_model=BlueprintRead, status_code=status.HTTP_201_CREATED)
def create_blueprint(
    payload: CreateBlueprintRequest,
    current_user: Annotated[User, Depends(require_user)],
    db: Annotated[Session, Depends(get_db)],
) -> GameBlueprint:
    game = _get_owned_game(payload.game_id, current_user, db)
    topics = list(
        db.scalars(
            select(Topic)
            .where(Topic.game_id == game.id)
            .order_by(Topic.sort_order.asc(), Topic.id.asc())
        )
    )
    if not topics:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Select or extract topics before creating a blueprint.")

    existing = db.scalar(select(GameBlueprint).where(GameBlueprint.game_id == game.id))
    if existing is not None and not payload.replace_existing:
        return existing
    if existing is not None:
        db.delete(existing)
        db.commit()

    generated = create_blueprint_from_topics(game, topics)
    blueprint = GameBlueprint(
        game_id=game.id,
        creator_id=current_user.id,
        title=generated.title,
        description=generated.description,
        total_levels=generated.total_levels,
        estimated_duration_minutes=generated.estimated_duration_minutes,
        topic_order=generated.topic_order,
        difficulty_distribution=generated.difficulty_distribution,
        reward_style=generated.reward_style,
        branching_style=generated.branching_style,
        approved=False,
    )
    db.add(blueprint)
    db.commit()
    db.refresh(blueprint)
    return blueprint


@blueprints_router.get("", response_model=BlueprintRead)
def get_blueprint(
    game_id: int,
    current_user: Annotated[User, Depends(require_user)],
    db: Annotated[Session, Depends(get_db)],
) -> GameBlueprint:
    return _get_owned_blueprint(game_id, current_user, db)


@blueprints_router.patch("", response_model=BlueprintRead)
def update_blueprint(
    game_id: int,
    payload: BlueprintUpdate,
    current_user: Annotated[User, Depends(require_user)],
    db: Annotated[Session, Depends(get_db)],
) -> GameBlueprint:
    blueprint = _get_owned_blueprint(game_id, current_user, db)
    for field, value in payload.model_dump(exclude_unset=True).items():
        setattr(blueprint, field, value)
    blueprint.approved = False
    db.add(blueprint)
    db.commit()
    db.refresh(blueprint)
    return blueprint


@blueprints_router.post("/approve", response_model=BlueprintRead)
def approve_blueprint(
    game_id: int,
    current_user: Annotated[User, Depends(require_user)],
    db: Annotated[Session, Depends(get_db)],
) -> GameBlueprint:
    blueprint = _get_owned_blueprint(game_id, current_user, db)
    blueprint.approved = True
    db.add(blueprint)
    db.commit()
    db.refresh(blueprint)
    return blueprint


def _get_owned_game(game_id: int, current_user: User, db: Session) -> Game:
    game = db.get(Game, game_id)
    if game is None or game.creator_id != current_user.id:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Game not found.")
    return game


def _get_owned_blueprint(game_id: int, current_user: User, db: Session) -> GameBlueprint:
    _get_owned_game(game_id, current_user, db)
    blueprint = db.scalar(select(GameBlueprint).where(GameBlueprint.game_id == game_id))
    if blueprint is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Blueprint not found.")
    return blueprint
